File: dump-possible-permissions.js
/*
dump-possible-permissions.js

Get all the permissions mentioned in WFActions.plist and save them to permissions.json.
All preexisting values in permissions.json are preserved, so permissions.json can be filled in manually.

This only works on macOS.
*/

const bplist = require('bplist-parser');
const { writeFileSync } = require('fs');

const permissionsJSONPath = './permissions.json';

let perms = [];

function pushOrNot(perm) {
    if (!perms.includes(perm)) {
        perms.push(perm)
    }
}

(async () => {
    const obj = await bplist.parseFile(require('./get-wfactions-path.js'))[0];

    for (const actionID in obj) {
        const resources  = obj[actionID].RequiredResources;
        if (resources) {
            for (const resource of resources) {
                if (typeof resource == "string") {
                    pushOrNot(resource);
                } else {
                    if (resource.WFAccountClass) {
                        pushOrNot(resource.WFAccountClass)
                    } else {
                        pushOrNot(resource.WFResourceClass);
                    }
                }
            }
        }
    }

    for (const perm of ["Apple TV Remote", "FaceTime", "Phone", "Run JavaScript", "Settings"]) {
        perms.push(perm)
    }

    perms.sort();

    const permsJSON = require(permissionsJSONPath);

    for (const p of perms) {
        if (permsJSON[p] === undefined) {
            permsJSON[p] = {
                icon: "",
                description: ""
            };
            console.log(`Permission '${p}' has no description`)
        }
    }

    writeFileSync(permissionsJSONPath, JSON.stringify(permsJSON, null, 4))
})();

File: dump_possible_permissions.py
import json
import plistlib

PERMISSIONS_JSON_PATH = "./permissions.json"
WFACTIONS_PLIST_PATH = '/System/Library/PrivateFrameworks/WorkflowKit.framework/Resources/WFActions.plist'

EXTRA_PERMS = ["Apple TV Remote", "FaceTime", "Phone", "Run JavaScript", "Settings"]

perms = []

def push_or_not(perm):
    """Accepts a string, determines whether to add it to the permissions list."""
    if not perm in perms:
        perms.append(perm)

with open(WFACTIONS_PLIST_PATH, "rb") as file:
    actions_dict = plistlib.load(file)

for action_id in actions_dict:
    action = actions_dict[action_id]
    if "RequiredResources" in action:
        resources = action["RequiredResources"]
        if resources:
            for resource in resources:
                if type(resource) is str:
                    push_or_not(resource)
                else:
                    if "WFAccountClass" in resource:
                        push_or_not(resource["WFAccountClass"])
                    else:
                        push_or_not(resource["WFResourceClass"])

for perm in EXTRA_PERMS:
    perms.append(perm)

perms.sort()

with open(PERMISSIONS_JSON_PATH, "r") as file:
    perms_json = json.load(file)

for perm in perms:
    if not perm in perms_json:
        perms_json[perm] = { "icon": "", "description": "" }
        print("Permission '" + perm + "' has no description")
    
with open(PERMISSIONS_JSON_PATH, "w") as file:
    json.dump(dict(perms_json), file, indent = 2)